### DISCUSSION TITLE
What would be the best way to remove all stale branches at one go

### DISCUSSION THREAD

**Discussion by sairohith1009** [community/community#54930] [New to GitHub]
Hi,

As we will have many stale branches in our GitHub repository. I want to delete all those branches as a part of GitHub Repo Clean-up.
What would be the best process to follow ?

Thanks in advance.

Regards,
Sai.

**Comment by NImaism** [comment]
Hi **sir**, you can easily find out which branches have not been merged into the main branch by using the following command `git branch --merged | grep -v "\*" | xargs -n 1 git branch -d`

**Comment by hrishikesh-kocharekar-dev** [comment]
Hi,
Review the branches: First, you'll want to review the branches in your repository to determine which ones are stale and no longer needed. You can use the "Branches" tab in your repository to view all branches, and sort by date or other criteria to help identify stale branches.
Delete branches: Once you've identified which branches to delete, you can delete them using the "Delete branch" button in the branch's details page. Alternatively, you can use a Git command such as `git branch -d <branch-name>` to delete branches locally and then push the changes to your remote repository.
Consider branch protection: If your repository is set up to require pull requests for merging changes, you may want to consider enabling branch protection rules to prevent accidental deletion of protected branches.

Schedule regular cleanups: To maintain a clean and organized repository, it's a good idea to schedule regular cleanups to remove stale branches. You can set up a reminder or create a recurring task to ensure that you perform cleanups on a regular basis.
It's important to note that deleting branches will permanently remove them from your repository, so be sure to double-check which branches you want to delete before proceeding. Additionally, if you have any continuous integration or deployment systems that depend on specific branches, be sure to update them accordingly.

**Comment by sairohith1009** [comment]
Is it possible to delete all the stale branches older than 30 days with one go ? i mean do we have any commands to do that ?

**Comment by ghostinhershell** [comment]
Hey @sairohith1009 and all 👋🏾 

I'm afraid we don't have a single API endpoint that you could query to show all stale branches for a given repository, similar to the [branch view list available through GitHub's web UI](https://docs.github.com/en/enterprise-cloud@latest/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/viewing-branches-in-your-repository). That being said, it would be possible to create a script to identify stale branches by querying several [GitHub REST API](https://docs.github.com/en/enterprise-server@3.12/rest?apiVersion=2022-11-28) endpoints. You may find the [official Octokit library or other third-party libraries](https://docs.github.com/en/enterprise-server@3.12/rest/using-the-rest-api/libraries-for-the-rest-api?apiVersion=2022-11-28) helpful with scripting this process. 
 
As a reminder, the concept of a stale branch only exists in the GitHub web application, it doesn't exist in Git. GitHub considers a branch as stale if no-one has committed to it in the last three months. With this in mind, you could the incorporate the following in your script:
 

- Query the [List branches](https://docs.github.com/en/enterprise-server@3.12/rest/branches/branches?apiVersion=2022-11-28#list-branches) REST API endpoint to obtain branches in the form of a JSON array object.
- Then loop over all branches in the array, extracting the `name` attribute.
- For each branch you could:
  - Call the [Get a branch](https://docs.github.com/en/enterprise-server@3.12/rest/branches/branches?apiVersion=2022-11-28#get-a-branch) REST API endpoint to obtain information about the specified branch.
  - From the `commit` object that contains information about the commit at the top of the branch, extract the `author date` and/or the `commit date`.
  - Perform a comparison on the chosen timestamp to identify if it is old enough to be considered "stale".
 
Please note that the [List branches](https://docs.github.com/en/enterprise-server@3.12/rest/branches/branches?apiVersion=2022-11-28#list-branches) endpoint is subject to [Pagination](https://docs.github.com/en/enterprise-server@3.12/rest/using-the-rest-api/using-pagination-in-the-rest-api?apiVersion=2022-11-28). By default, this endpoint will return 30 results but can be changed by including the `per_page=NUMBER` parameter in the request.
 
I hope this helps. If you have any further questions about this, please let us know.
